crates/nu-parser/src/parser.rs

```diff
@@ -2834,12 +2834,61 @@ pub fn parse_string_strict(
     }
 }
 
+fn parse_list_shape(
+    working_set: &StateWorkingSet,
+    bytes: &[u8],
+    span: Span,
+) -> (SyntaxShape, Option<ParseError>) {
+    assert!(bytes.starts_with(b"list<"));
+
+    let start = span.start + 5;
+    let end = span.end;
+    let is_terminated = bytes.ends_with(b">");
+    let unterminated_error = ParseError::LabeledError(
+        "Unterminated list type annotation".into(),
+        "use `>` to terminate it".into(),
+        span,
+    );
+
+    let inner_span = Span::new(start, if is_terminated { end - 1 } else { end });
+
+    let src = working_set.get_span_contents(inner_span);
+
+    // list<>
+    if src.is_empty() && is_terminated {
+        return (SyntaxShape::List(Box::new(SyntaxShape::Any)), None);
+    } else if src.is_empty() {
+        // list<
+        return (
+            SyntaxShape::List(Box::new(SyntaxShape::Any)),
+            Some(unterminated_error),
+        );
+    }
+
+    let (shape, err) = parse_shape_name(working_set, src, inner_span);
+    if err.is_some() {
+        return (SyntaxShape::List(Box::new(shape)), err);
+    }
+
+    let error = if !is_terminated {
+        Some(unterminated_error)
+    } else {
+        None
+    };
+
+    (SyntaxShape::List(Box::new(shape)), error)
```

Comment:
All those errors still return a `SyntaxShape::List()`. Is this desirable or should we reject bogus type names in there outright? (Haven't checked how the parse/error pair is handled upstream in this case)

Reply:
Not a parser expert, seems to work in the tests, which is the important bit.
If anyone of the parser experts wants to weigh in, would be welcome.